Automattic/newspack-newsletters
# feat: woocommerce data sync tools

### All Submissions:






### Changes proposed in this Pull Request:

Moves the logic for syncing Woo customers and orders to the ESP from the `newspack-subscription-migrations` plugin to here.

The original class was refactored so it provides generic methods to be extended by the existing CLI and new admin functionality.

Some changes to the generic logic are worth mentioning:

 - `can_sync_contacts( bool $return_errors )` was implemented so it can be called across CLI and admin functionality with proper error handling
 - `get_batch_of_customers()` was changed to `get_batch_of_readers()`. It's now using `\Newspack\Reader_Activation::get_reader_roles()` because although they are customers, every WP user can be a customer and the goal is to sync RAS readers
 - The private `sync( $contact )` method (previously `sync_contact()`) was modified to support any ESP through the `Newspack_Newsletters_Contacts::upsert()` method. For that, [a new helper method is proposed for RAS](https://github.com/Automattic/newspack-plugin/pull/3355) so we can get the master list given any provider

The CLI command, originally `wp newspack woo resync` is now `newspack-newsletters woo resync` but preserves all the original functionality and args support. The exception is for the flag `--migrated-subscriptions`, which requires the `newspack-subscription-migrations` plugin to be installed, otherwise it returns an error.

The new functionality allows an admin user to sync contacts through the dashboard user table, both individually and through the bulk action dropdown:

| Single | Bulk |
| --- | --- |
| <img width="376" alt="image" src="https://github.com/user-attachments/assets/8ce4e39a-42f7-460b-99ec-46c778715560"> | <img width="255" alt="image" src="https://github.com/user-attachments/assets/07a46a1c-db4e-4f7e-b8b4-19a850e45294"> |

### How to test the changes in this Pull Request:

1. Check out this PR and https://github.com/Automattic/newspack-plugin/pull/3355

#### CLI

1. Run `wp newspack-newsletters woo resync --dry-run` for a full dry run
2. Confirm the command runs successfully, going through all the readers in your database
3. Choose a reader and run `wp newspack-newsletters woo resync --user-ids=USER_ID` with their user ID
4. Confirm via PHP logs that `[Newspack_Newsletters_Contacts::upsert]: Adding contact to list(s): LIST_ID. Provider is PROVIDER.` is called and you can inspect the normalized contact data
5. Do the same but now with 2 user IDs: `wp newspack-newsletters woo resync --user-ids=USER_ID1,USER_ID2`
6. Confirm via PHP logs the 2 entries as in step 4
7. Grab 2 Woo Subscription IDs and run: `wp newspack-newsletters woo resync --subscription-ids=ID1,ID2`
8. Confirm the command runs without issues and the PHP log entries for the `::upsert()` is correct
9. Grab 2 Woo Order IDs and run: `wp newspack-newsletters woo resync --order-ids=ID1,ID2`
10. Confirm the command runs without issues and the PHP log entries for the `::upsert()` is correct
11. Cancel a reader subscription and run a sync for active subs only, along with a reader with an active subscription: `wp newspack-newsletters woo resync --active-only --user-ids=USER_ID1,USER_ID2`
12. Confirm only the reader with an active sub got synced
13. Test dry run syncs modifying `--offset`, `--max-batches`, and `--batch-size` to confirm these values are applied correctly

#### Admin

1. Navigate to "Users", in WP dashboard
2. Hover the rows and confirm you see a "Resync contact to ESP" row action link
3. Click and confirm you get a success message and the PHP logs show `::upsert()` was called correctly
4. Select multiple readers, click on the Bulk actions dropdown, select "Resync to the ESP", and apply
5. Confirm you get a success message and the PHP logs show `::upsert()` was called correctly for every selected contact

### Other information:
